File: csharp/EviFluor/singlemeasurement.cs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

using System;
using System.Text.Json.Nodes;

namespace Hse.EviFluor;

/// <summary>
/// Represents a single measurement consisting the 470 nm wavelength channel.
/// </summary>
public class SingleMeasurement
{
    /// <summary>
    /// Gets or sets the channel for 470 nm measurements.
    /// </summary>
    public Channel Channel470 { get; set; }

    /// <summary>
    /// Initializes a new instance of the <see cref="SingleMeasurement"/> class with optional channels.
    /// </summary>
    /// <param name="channel470">The channel for 470 nm (default: new Channel()).</param>
    public SingleMeasurement(Channel? channel470 = null)
    {
        Channel470 = channel470 ?? new Channel();
    }

    /// <summary>
    /// Returns a string representation of the measurement values.
    /// </summary>
    /// <returns>A formatted string containing all channel values.</returns>
    public override string ToString()
    {
        return $"470: [{Channel470}]]";
    }

    /// <summary>
    /// Converts the measurement data to a JSON representation.
    /// </summary>
    /// <returns>A JsonNode representing the measurement.</returns>
    public JsonNode ToJson()
    {
        return Channel470.ToJson();
    }


    /// <summary>
    /// Returns the difference between sample and dark.self->
    /// </summary>
    /// <returns>Difference between sample and dark</returns>
    public double Delta()
    {
        return Channel470.Delta();
    }

    /// <summary>
    /// Creates a <see cref="SingleMeasurement"/> instance from a JSON node.
    /// </summary>
    /// <param name="node">The JSON node containing measurement data.</param>
    /// <returns>A new <see cref="SingleMeasurement"/> instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if the JSON node is null or missing required properties.</exception>
    public static SingleMeasurement FromJson(JsonNode node)
    {
        if (node == null) throw new ArgumentNullException(nameof(node));

        return new SingleMeasurement(Channel.FromJson(node));
    }
}

File: python/hse/evifluor/singlemeasurement.py
# SPDX-License-Identifier: MIT
# SPDX-FileCopyrightText: © 2025 HSE AG, <opensource@hseag.com>

from hse.evifluor.channel import Channel


class SingleMeasurement:
    """Represents a single measurement consisting of the 470 nm wavelength channel."""

    def __init__(self, channel_470 = Channel()):
        """Initializes the measurement with an optional 470 nm channel."""
        self.channel_470 = channel_470
        
    def __repr__(self):
        """Returns a compact textual representation of the measurement."""
        return "470: [{}]".format(self.channel_470)
        
    def to_json(self):
        """Converts the measurement data to a JSON representation."""
        return self.channel_470.to_json()
        
    def delta(self):
        """Returns the difference between sample signal and dark signal."""
        return self.channel_470.delta()
               
    @staticmethod
    def from_json(node):
        """Creates a SingleMeasurement instance from a JSON node."""
        return SingleMeasurement(Channel.from_json(node))
